CORE-13739 Update schema registry to use chunked_vector

nguyen-andrew opened the pull request:
This PR addresses oversized allocations in schema registry. A recent instance of this issue occurred from several upserts to the same subject, causing its vector to grow very large. This issue is addressed by replacing schema registry's internal `std::vector` containers with `chunked_vector` to avoid large contiguous allocations.

## Backports Required

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.2.x
- [x] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=28073) <br/>All committers have signed the CLA.

dotnwat commented on src/v/pandaproxy/schema_registry/store.h:
```diff
@@ -833,7 +833,7 @@ class store {
         }
     };
-    using schema_map = absl::btree_map<schema_id, schema_entry>;
+    using schema_map = chunked_hash_map<schema_id, schema_entry>;
```
btree was already a chunked data structure, but this change would cause container iteration to become unordered. is that ok?

nguyen-andrew replied:
Ah right, I was too overenthusiastic with chunking things. There looks like there is some code that relies on the ordering. I'll undo this, thanks Noah!

dotnwat commented on src/v/pandaproxy/schema_registry/store.h:
```diff
@@ -860,18 +867,29 @@ class store {
+        if (!inc_del && v_it->deleted) {
+            return not_found(sub_entry.first, version);
+        }
+        return v_it;
```
its not really clear how this change relates to changing the container to chunked vector.

nguyen-andrew replied:
I temporarily duplicated the logic from the overload below and placed it here. Did this because it doesn't seem like `detail::make_non_const_iterator` and its `container.erase(it, it)` trick (TIL!) works when `sub_entry.second.versions` is made into a `chunked_vector` (since there's no erase function available).

dotnwat commented on src/v/pandaproxy/schema_registry/store.h:
```diff
@@ -684,7 +687,11 @@ class store {
+            // emplace at the back and rotate it into position
+            auto idx = v_it - versions.begin();
+            versions.emplace_back(version, id, deleted);
+            std::rotate(versions.begin() + idx, versions.end() - 1, versions.end());
```
just thinking outloud: if we are changing to a chunked data structure because they were getting too large, then maybe we aren't using the best data structure if we have to jump through these std::rotate hoops (or maybe we should add better interfces to chunked vector)

michael-redpanda commented:
/dt

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--82c203ad-ae2d-40c0-a83d-8ae9ca03e3fd-->
<!--TEST RESULTS-->
<!--22c172dd-c6b1-4880-b62c-ac88140c96a1-->
<!--TEST RESULTS-->
<!--ca953a8c-a9ae-482d-b23c-735d282141d5-->
<!--TEST RESULTS-->
<!--237b67da-a1a3-42b2-a4c8-ddce8ce8304f-->
<!--TEST RESULTS-->
<!--2184f0ff-9c71-45d2-b0dd-5d68989fbb18-->
<!--TEST RESULTS-->
<!--f196c274-887c-48b1-90f5-1eb47d014954-->
<!--TEST RESULTS-->
<!--f4148c06-ceb9-4ffe-97a4-4c473c287170-->
<!--TEST RESULTS-->
<!--5802c544-1897-485e-9429-4d23b9821197-->
<!--TEST RESULTS-->
<!--96912e40-1277-451d-b1e4-18623afa4a3e-->
<!--TEST RESULTS-->
<!--76de02fb-7808-4d4f-b895-5e20e1526ecc-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74323">build#74323</a></summary>


| test_class                          | test_method                                             | test_arguments                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                                   |
|:------------------------------------|:--------------------------------------------------------|:---------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodesDecommissioningTest            | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": false, "delete_topic": true} | integration | https://buildkite.com/redpanda/redpanda/builds/74323#0199ed93-fd8e-4dd5-81fd-81d51676f78e | FLAKY         | 18/21    | upstream reliability is '97.87234042553192'. current run reliability is '85.71428571428571'. drift is 12.15805 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
| NodesDecommissioningTest            | test_recommissioning_node                               | {"cloud_topic": true}                        | integration | https://buildkite.com/redpanda/redpanda/builds/74323#0199ed93-fd8b-43e8-b4d9-0380417539cb | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_recommissioning_node                               |
| WriteCachingFailureInjectionE2ETest | test_crash_all                                          | {"use_transactions": false}                  | integration | https://buildkite.com/redpanda/redpanda/builds/74323#0199ed93-fd8a-4771-966e-62b3b13ede20 | FLAKY         | 14/21    | upstream reliability is '81.68908819133034'. current run reliability is '66.66666666666666'. drift is 15.02242 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all                               |
| WriteCachingFailureInjectionTest    | test_unavoidable_data_loss                              | null                                         | integration | https://buildkite.com/redpanda/redpanda/builds/74323#0199ed93-fd90-4e37-8264-4bb46b0053d1 | FLAKY         | 17/21    | upstream reliability is '90.68100358422939'. current run reliability is '80.95238095238095'. drift is 9.72862 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss                      |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74542">build#74542</a></summary>


| test_class                    | test_method                          | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                               |
|:------------------------------|:-------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_replication_with_failures       | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74542#019a02e3-bfb7-42c6-85a4-9d171d55eb59 | FLAKY         | 10/21    | upstream reliability is '63.13173156586578'. current run reliability is '47.61904761904761'. drift is 15.51268 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures |
| ShadowLinkingReplicationTests | test_replication_with_failures       | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74542#019a02ec-561c-437e-966b-c8d76b5821c4 | FLAKY         | 8/21     | upstream reliability is '63.13173156586578'. current run reliability is '38.095238095238095'. drift is 25.03649 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures |
| DataMigrationsApiTest         | test_creating_and_listing_migrations | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74542#019a02e3-bfb4-4739-978a-69a6fac2bfd1 | FLAKY         | 19/21    | upstream reliability is '99.00990099009901'. current run reliability is '90.47619047619048'. drift is 8.53371 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations   |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74564">build#74564</a></summary>


| test_class                        | test_method                        | test_arguments             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                |
|:----------------------------------|:-----------------------------------|:---------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| LeaderEpochTest                   | TestGetLeaderEpochWhileReconciling |                            | unit        | https://buildkite.com/redpanda/redpanda/builds/74564#019a038d-085a-4423-810d-bed8d0739d83 | FAIL          | 0/1      |                                                                                                                                                                        |                                                                                                                                                                             |
| VerifyConsumerOffsetsThruUpgrades | test_consumer_group_offsets        | {"versions_to_upgrade": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/74564#019a03c3-6cf5-4c1d-a745-fec11e3b4769 | FLAKY         | 20/21    | upstream reliability is '99.77220956719817'. current run reliability is '95.23809523809523'. drift is 4.53411 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=VerifyConsumerOffsetsThruUpgrades&test_method=test_consumer_group_offsets |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74646">build#74646</a></summary>


| test_class                 | test_method          | test_arguments      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                  |
|:---------------------------|:---------------------|:--------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| QuickTerminateTest         | test_terminate       | null                | integration | https://buildkite.com/redpanda/redpanda/builds/74646#019a091a-10e7-4cea-9ab4-805d800b511b | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=QuickTerminateTest&test_method=test_terminate               |
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74646#019a091a-10ea-43a6-a546-c048b531090a | FLAKY         | 9/21     | upstream reliability is '39.75155279503105'. current run reliability is '42.857142857142854'. drift is -3.10559 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/74646#019a091a-10dc-4800-86a7-4d5eed832b1d | FLAKY         | 4/21     | upstream reliability is '65.9090909090909'. current run reliability is '19.047619047619047'. drift is 46.86147 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/74646#019a0920-2f82-4d39-a869-fe4abee00d3e | FLAKY         | 2/21     | upstream reliability is '57.99999999999999'. current run reliability is '9.523809523809524'. drift is 48.47619 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74830">build#74830</a></summary>


| test_class               | test_method                          | test_arguments        | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                              |
|:-------------------------|:-------------------------------------|:----------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest    | test_creating_and_listing_migrations | null                  | integration | https://buildkite.com/redpanda/redpanda/builds/74830#019a128c-8a4a-41ba-9cc3-e63d86a84c19 | FLAKY         | 18/21    | upstream reliability is '99.49324324324324'. current run reliability is '85.71428571428571'. drift is 13.77896 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations  |
| NodesDecommissioningTest | test_recommissioning_node_finishes   | {"cloud_topic": true} | integration | https://buildkite.com/redpanda/redpanda/builds/74830#019a128c-8a42-463b-a6b9-c71f7cbcce56 | FLAKY         | 20/21    | upstream reliability is '98.17629179331307'. current run reliability is '95.23809523809523'. drift is 2.9382 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_recommissioning_node_finishes |
| PartitionMovementTest    | test_static                          | {"num_to_upgrade": 0} | integration | https://buildkite.com/redpanda/redpanda/builds/74830#019a128c-8a42-463b-a6b9-c71f7cbcce56 | FLAKY         | 17/21    | upstream reliability is '92.4496644295302'. current run reliability is '80.95238095238095'. drift is 11.49728 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_static                           |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74857">build#74857</a></summary>


| test_class                      | test_method                        | test_arguments         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                              |
|:--------------------------------|:-----------------------------------|:-----------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AdminV2ListKafkaConnectionsTest | test_list_kafka_connections        | null                   | integration | https://buildkite.com/redpanda/redpanda/builds/74857#019a1362-5c2a-4107-be27-13f6eede2d25 | FLAKY         | 19/21    | upstream reliability is '99.57446808510639'. current run reliability is '90.47619047619048'. drift is 9.09828 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AdminV2ListKafkaConnectionsTest&test_method=test_list_kafka_connections |
| NodesDecommissioningTest        | test_recommissioning_node_finishes | {"cloud_topic": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74857#019a136b-5ea2-4199-81d5-769cc5ce8146 | FLAKY         | 18/21    | upstream reliability is '98.39142091152814'. current run reliability is '85.71428571428571'. drift is 12.67714 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_recommissioning_node_finishes |
| PartitionMovementTest           | test_static                        | {"num_to_upgrade": 0}  | integration | https://buildkite.com/redpanda/redpanda/builds/74857#019a1362-5c2d-4377-ae9e-7a47e271f3a7 | FLAKY         | 16/21    | upstream reliability is '91.92245557350566'. current run reliability is '76.19047619047619'. drift is 15.73198 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_static                           |
| ShadowLinkingRandomOpsTest      | test_node_operations               | {"failures": false}    | integration | https://buildkite.com/redpanda/redpanda/builds/74857#019a1362-5c2e-4982-b647-8352c622e74d | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations             |
| ShadowLinkingRandomOpsTest      | test_node_operations               | {"failures": false}    | integration | https://buildkite.com/redpanda/redpanda/builds/74857#019a136b-5ea5-48fd-b426-51ab62dd4a52 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations             |
| ShadowLinkingRandomOpsTest      | test_node_operations               | {"failures": true}     | integration | https://buildkite.com/redpanda/redpanda/builds/74857#019a1362-5c2f-4f87-90dc-590e430e372e | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations             |
| ShadowLinkingRandomOpsTest      | test_node_operations               | {"failures": true}     | integration | https://buildkite.com/redpanda/redpanda/builds/74857#019a136b-5ea6-4739-afb6-fe3589edbd89 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations             |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74916">build#74916</a></summary>


| test_class                          | test_method                         | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   | test_history   |
|:------------------------------------|:------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|:---------------|
| src/v/cluster_link/tests/utils_test | src/v/cluster_link/tests/utils_test |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/74916#019a1678-a398-4022-b8fb-aaeefe80bf7d | FAIL          | 0/1      |          |                |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74967">build#74967</a></summary>


| test_class               | test_method                                             | test_arguments                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                                   |
|:-------------------------|:--------------------------------------------------------|:---------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AuditLogTestKafkaApi     | test_no_auth_enabled                                    | {"audit_transport_mode": "rpc"}              | integration | https://buildkite.com/redpanda/redpanda/builds/74967#019a17c5-8c25-4818-81f7-fa8c0115cfa1 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AuditLogTestKafkaApi&test_method=test_no_auth_enabled                                        |
| SegmentMsTest            | test_segment_rolling_with_retention_consumer            | null                                         | integration | https://buildkite.com/redpanda/redpanda/builds/74967#019a17c5-8c28-49f5-acde-06da7ebfea69 | FLAKY         | 19/21    | upstream reliability is '92.77721261444557'. current run reliability is '90.47619047619048'. drift is 2.30102 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer                       |
| NodesDecommissioningTest | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": false, "delete_topic": true} | integration | https://buildkite.com/redpanda/redpanda/builds/74967#019a17c5-8c1e-48e7-a173-9f91ee99fdd3 | FLAKY         | 20/21    | upstream reliability is '98.21428571428571'. current run reliability is '95.23809523809523'. drift is 2.97619 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75075">build#75075</a></summary>


| test_class                                | test_method                   | test_arguments                                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                          |
|:------------------------------------------|:------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest                 | test_write                    | null                                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/75075#019a2834-62e7-4e7a-9ed7-761033e0792a | FLAKY         | 18/21    | upstream reliability is '96.16724738675958'. current run reliability is '85.71428571428571'. drift is 10.45296 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTxTest&test_method=test_write                                    |
| EndToEndShadowIndexingTestWithDisruptions | test_write_with_node_failures | {"cloud_storage_type": 2}                                                                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/75075#019a2834-62e6-4e7d-945a-6a0f3e39f516 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndShadowIndexingTestWithDisruptions&test_method=test_write_with_node_failures |
| TieredStorageTest                         | test_tiered_storage           | {"cloud_storage_type_and_url_style": [1, "path"], "test_case": {"name": "(TS_Read == True, TS_Timequery == True, SpilloverManifestUploaded == True)"}} | integration | https://buildkite.com/redpanda/redpanda/builds/75075#019a2834-62e6-4e7d-945a-6a0f3e39f516 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TieredStorageTest&test_method=test_tiered_storage                                   |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75322">build#75322</a></summary>


| test_class                          | test_method                                | test_arguments                                                                | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                                               |
|:------------------------------------|:-------------------------------------------|:------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DeleteRecordsTest                   | test_delete_records_concurrent_truncations | {"cloud_storage_enabled": true, "truncate_point": "one_below_high_watermark"} | integration | https://buildkite.com/redpanda/redpanda/builds/75322#019a358b-22ff-44b1-a2c7-afdfd8129694 | FLAKY         | 15/21    | upstream reliability is '94.15204678362574'. current run reliability is '71.42857142857143'. drift is 22.72348 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DeleteRecordsTest&test_method=test_delete_records_concurrent_truncations                 |
| LogCompactionTxRemovalUpgradeTest   | test_tx_control_batch_removal_with_upgrade | {"test_case_name": "Mixed aborts and commits"}                                | integration | https://buildkite.com/redpanda/redpanda/builds/75322#019a3592-382e-4389-a9e2-169124286924 | FLAKY         | 19/21    | upstream reliability is '97.44680851063829'. current run reliability is '90.47619047619048'. drift is 6.97062 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalUpgradeTest&test_method=test_tx_control_batch_removal_with_upgrade |
| RedpandaNodeOperationsSmokeTest     | test_node_ops_smoke_test                   | {"cloud_storage_type": 1, "mixed_versions": true}                             | integration | https://buildkite.com/redpanda/redpanda/builds/75322#019a358b-2312-4c4a-bdd6-1ecf61133866 | FLAKY         | 20/21    | upstream reliability is '99.85835694050992'. current run reliability is '95.23809523809523'. drift is 4.62026 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test                     |
| ScalingUpTest                       | test_fast_node_addition                    | null                                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/75322#019a3592-382d-450f-81ab-e61df840253b | FLAKY         | 20/21    | upstream reliability is '99.581589958159'. current run reliability is '95.23809523809523'. drift is 4.34349 and the allowed drift is set to 50. The test should PASS     | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ScalingUpTest&test_method=test_fast_node_addition                                        |
| WriteCachingFailureInjectionE2ETest | test_crash_all                             | {"use_transactions": false}                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/75322#019a3592-382b-47e3-87b2-fe35daf5febe | FLAKY         | 20/21    | upstream reliability is '84.54481298517996'. current run reliability is '95.23809523809523'. drift is -10.69328 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all                           |
</details>
<!--/body-->

nguyen-andrew committed:
sr: ducktape test for oversized allocs for subject_entry::versions

Adding an ignored ducktape test that causes an oversized allocation
warning by expanding the subject_entry::versions vector.

nguyen-andrew committed:
sr: update subject_entry::versions to use chunked_vector

Keeping the related ducktape test parameter (4097) ignored
since it takes more time to run, but it has been manually
verified as passing.

nguyen-andrew committed:
sr: removing unused subject_version_entry::written_at vector

Discovered that subject_version_entry::written_at field was
not actually being used anywhere. Instead of converting it
to a chunked_vector and having to add explicit copies to
various interactions with subject_version_entry, we can
simply remove the unused/empty written_at field.

nguyen-andrew committed:
sr: ducktape test for oversized allocs for subject_entry::written_at

Adding an ignored ducktape test that causes an oversized allocation
warning by expanding the subject_entry::written_at vector.

nguyen-andrew committed:
sr: update subject_entry::written_at to use chunked_vector

Also enables the ducktape test that validates that the changes
no longer causes an oversized alloc warning.

nguyen-andrew committed:
sr: ducktape test for oversized allocs for store::get_version_ids

Adding an ignored ducktape test that causes an oversized allocation
warning by expanding the res vector in store::get_version_ids.

nguyen-andrew committed:
sr: update get_version_ids to use and return a chunked_vector

Update store/sharded_store::get_version_ids to use and return
a chunked vector. This fixes an oversized allocation case
triggered by posting many unique versions to the same subject.

nguyen-andrew committed:
sr: ducktape test for oversized allocs for store::get_versions

Extending ducktape test with an additional ignored parameterization
to exercise oversized allocations in store::get_versions.

nguyen-andrew committed:
sr: update get_versions to return a chunked_vector

Update store/sharded_store::get_versions to return a
chunked vector. This fixes an oversized allocation case
triggered by getting the subject versions for a subject
with many unique versions.

nguyen-andrew committed:
sr: ducktape test for oversized allocs in delete subject

Extending ducktape test to delete a subject with several
unique versions to exercise oversized allocatoins in
the delete subject path in seq_writer.

nguyen-andrew committed:
sr: update delete subject to use chunked_vector

nguyen-andrew committed:
sr: update delete in seq_writer to avoid oversized allocations

Update seq_writer to better handle impermanent deletions of subjects.
Previously, deleting a subject with many versions causes seq_writer
to call parallel_transforms and create a task for each version (to
check if any of the subject versions are being referenced); for
thousands of versions, this can cause the underlying task queue
to enlarge and trigger its own oversized allocation warning. This
modified implementation simplifies the code and makes it more
efficient to check if a subject is referenced.

nguyen-andrew committed:
sr: ducktape test for oversized allocs in delete mode

Extending ducktape test to delete a subject mode with
several seq_markers to exercise oversized allocations.

nguyen-andrew committed:
sr: update get_subject_mode_written_at to avoid oversized allocs

Update store/sharded_store::get_subject_mode_written_at to use
a chunked_vector. This fixes an oversized allocation case
triggered by several mode changes followed by a delete.

nguyen-andrew committed:
sr: ducktape test for oversized allocs in delete subject compatibility

Extending ducktape test to delete a subject compatibility with several
seq_markers to exercise oversized allocations.

nguyen-andrew committed:
sr: update get_subject_config_written_at to avoid oversized allocs

Update store/sharded_store::get_subject_config_written_at to use
a chunked_vector. This fixes an oversized allocation case triggered
by several config changes followed by a delete.

nguyen-andrew committed:
sr: update get_subject_version_written_at to use chunked_vector

nguyen-andrew committed:
sr: ducktape test for oversized allocs in sharded_store::referenced_by

Adding ducktape test creating many subjects referencing the same base
subject to trigger an oversized allocation warning. Setting the test as
ignored because it takes over a minute to run.

nguyen-andrew committed:
sr: update sharded_store::referenced_by to avoid oversized allocs

Update sharded_store::referenced_by to use a chunked_vector. This
fixes an oversized allocation case when a subject is referenced by
many other subjects.

nguyen-andrew committed:
sr: update compatability checks to use chunked_vectors

nguyen-andrew committed:
sr: use chunked_vector for schema_definition::references

nguyen-andrew committed:
sr: use chunked_vector for raw_compatibility_result::_errors

nguyen-andrew commented:
Force pushed to branch - I restructured the commits to better organize the changes and introduce some tests. Also rebased on the latest dev. Some prior comments may now be outdated, so please review the latest diff.

nguyen-andrew commented on src/v/pandaproxy/schema_registry/seq_writer.cc:
```diff
@@ -524,17 +526,28 @@ seq_writer::do_delete_subject_impermanent(subject sub, model::offset write_at) {
-        throw as_exception(has_references(sub, versions.back()));
+    // Check for references. Done in batches to prevent
+    // oversized allocations in seastar's task queue
+    constexpr size_t batch_size = 1000;
```
Drawing attention to this change.

The original code seems to create a parallel task for each version. But if the subject being deleted has several versions (e.g. thousands), it seems to blow up its task queue (as indicated by the log message: `"Too long queue accumulated for main"`) and cause an oversized alloc for the underlying task queue circular buffer.

nguyen-andrew commented on src/v/pandaproxy/schema_registry/types.h:
```diff
@@ -205,14 +206,16 @@ class schema_definition {
 
     auto destructure() && {
-        return make_tuple(std::move(_def), _type, std::move(_refs));
+        return std::make_tuple(std::move(_def), _type, std::move(_refs));
```
Need to use `std::make_tuple` for the code to compile.

nguyen-andrew committed:
sr: separated oversized alloc ducktape tests as scale tests.

Moved the oversized alloc ducktape tests to be separate scale tests
and annotated them with @skip_debug_mode (since oversized alloc
warnings only occur in release). Added types for CI type checking.

nguyen-andrew commented:
Force pushed:
- Swapped chunked implementation of `parallel_transform` with `ss::max_concurrent_for_each` instead.
- Removed unused/ignored tests
- Rebased on dev

BenPope commented on src/v/pandaproxy/schema_registry/store.h:
```diff
@@ -503,12 +503,10 @@ class store {
+        markers.erase_to_end(
+          std::ranges::remove_if(markers, [&version](auto sm) {
+              return sm.version == version;
+          }).begin());
```
nitpick: You could use a projection
```suggestion
          std::ranges::remove(markers, version, &seq_marker::version).begin());
```

BenPope commented on src/v/pandaproxy/schema_registry/seq_writer.cc:
```diff
@@ -94,7 +94,9 @@ struct batch_builder : public storage::record_batch_builder {
-    void operator()(const std::vector<seq_marker>& sequences) {
+    template<typename Container>
+    requires std::same_as<std::ranges::range_value_t<Container>, seq_marker>
+    void operator()(const Container& sequences) {
```
Any reason not to be specific?
```suggestion
    void operator()(const chunked_vector<seq_marker>& sequences) {
```

nguyen-andrew replied:
I think I was making the `chunked_vector` changes incrementally, so I had made this generic, but I can change it to be specific now.

BenPope commented on src/v/pandaproxy/schema_registry/store.h:
```diff
@@ -497,7 +497,10 @@ class store {
         }
 
-        versions.erase(v_it);
+        // chunked_vector doesn't support erase(), so we need to
+        // manually erase
+        std::shift_left(v_it, versions.end(), 1);
+        versions.pop_back();
```
It's tempting to implement `chunked_vector::erase(iterator)`

nguyen-andrew replied:
Yea, might be worth tackling in another PR?

BenPope commented on src/v/pandaproxy/schema_registry/store.h:
```diff
@@ -684,7 +687,12 @@ class store {
             *v_it = subject_version_entry(version, id, deleted);
         } else {
-            versions.emplace(v_it, version, id, deleted);
+            // chunked_vector doesn't support emplace(), so we need to manually
+            // emplace at the back and rotate it into position
```
Again, it might be worth implementing `chunked_vector::emplace(iterator, Args&&... args)`.

nguyen-andrew replied:
Worth tackling in another PR?

BenPope commented on src/v/pandaproxy/schema_registry/seq_writer.cc:
```diff
@@ -529,10 +529,18 @@ seq_writer::do_delete_subject_impermanent(subject sub, model::offset write_at) {
+      batch_size,
+      [this, &sub, &is_referenced](const auto& ver) {
+          return _store.is_referenced(sub, ver).then(
+            [&is_referenced](bool b) { is_referenced.push_back(b); });
```
This whole algorithm is horribly inefficient (probably my fault):
1. We don't need a `vector<bool>`, a single `bool` will suffice, eliminating the later `any_of`.
2. Every version does a cross-shard call to the owner of the subject, followed by a many cross-shard calls in the map-reduce.

This can be vastly improved by implementing an `is_referenced` that operates on the entire subject:
```c++
ss::future<bool> sharded_store::is_referenced(subject sub) {
    // Find all the schema that reference this subject
    auto references = co_await _store.map_reduce0(
      [sub{std::move(sub)}](store& s) {
          return s.referenced_by(sub);
      },
      store::schema_id_set{},
      set_accumulator);
    // Find whether any subject version reference any of the schema
    co_return co_await _store.map_reduce0(
      [refs{std::move(references)}](store& s) {
          return s.subject_versions_has_any_of(refs, include_deleted::no);
      },
      false,
      std::logical_or<>{});
}
```

BenPope commented on tests/rptest/tests/schema_registry_test.py:
```diff
@@ -4739,6 +4739,35 @@ def test_id_lookup_multiple_matches(self):
+    @cluster(num_nodes=1)
+    @ignore(iterations=4097)
+    @parametrize(
+        iterations=4097
```
"several" :rofl:

BenPope commented on src/v/pandaproxy/schema_registry/types.h:
```diff
@@ -277,6 +279,10 @@ class protobuf_schema_definition {
 
     constexpr schema_type type() const { return schema_type::protobuf; }
 
+    protobuf_schema_definition copy() const {
+        return protobuf_schema_definition{_impl, _refs.copy()};
+    }
```
Is this used anywhere?

nguyen-andrew replied:
Yea I needed to add it for this: https://github.com/redpanda-data/redpanda/pull/28073/files#diff-18b29034a28d3fd2d4fb120c3fa1d9d86bf92aaf8c39c8740894b71413dc10caR122

BenPope commented on tests/rptest/tests/schema_registry_test.py:
```diff
@@ -29,7 +29,7 @@
 )
 from confluent_kafka.serialization import MessageField, SerializationContext
-from ducktape.mark import matrix, parametrize, ignore
+from ducktape.mark import matrix, parametrize
```
It's possible for things to regress, we should keep these tests in some form.

You can mark them `@skip_debug_mode`, as an oversize alloc is not reported in debug mode, and maybe move them to `rptest/scale_tests`.

nguyen-andrew replied:
Good idea, moved

BenPope reviewed (commented):
This looks great.

nguyen-andrew commented:
Force push to address PR comments, notably:
- Updated `seq_writer::do_delete_subject_impermanent` to replace usage of `ssx::parallel_transform`/`ss::max_concurrent_for_each` with a more simple approach.
- Moved ducktape tests to scale_tests & annotated with @skip_debug_mode

BenPope commented on tests/rptest/scale_tests/schema_registry_scale_test.py:
```diff
@@ -0,0 +1,218 @@
+# Copyright 2021 Redpanda Data, Inc.
```
```suggestion
# Copyright 2025 Redpanda Data, Inc.

BenPope commented:
/cdt
dt-repeat=10
tests/rptest/scale_tests/schema_registry_scale_test.py

BenPope commented:
/cdt
dt-repeat=10
skip-redpanda-build
release
tests/rptest/scale_tests/schema_registry_scale_test.py

nguyen-andrew commented:
/cdt
dt-repeat=10
skip-redpanda-build
release
tests/rptest/scale_tests/schema_registry_scale_test.py

rpdevmp commented:
/cdt
dt-repeat=1
release
tests/rptest/scale_tests/schema_registry_scale_test.py

BenPope commented on tests/rptest/scale_tests/schema_registry_scale_test.py:
```diff
@@ -0,0 +1,216 @@
+    @cluster(num_nodes=1)
+    @skip_debug_mode
+    @parametrize(iterations=2049)  # oversized alloc for subject_entry::written_at.
+    @parametrize(iterations=4097)  # oversized alloc for subject_entry::versions.
```
We don't need both of these.

BenPope commented on tests/rptest/scale_tests/schema_registry_scale_test.py:
```diff
@@ -0,0 +1,216 @@
+    @cluster(num_nodes=1)
+    @skip_debug_mode
+    @parametrize(iterations=8194)  # oversized alloc for store::get_version_ids.
+    @parametrize(iterations=32769)  # oversized alloc for store::get_versions.
```
We don't need both of these.

nguyen-andrew commented:
/dt
dt-repeat=10
tests/rptest/scale_tests/schema_registry_scale_test.py

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74857-->
<!--f7b95999-b3ab-4950-9ad7-2d8e3ae47878-->
<!--4606adc8-0f1c-487d-9077-a61d87bd52e3-->
<!--defd359a-3965-4751-aecc-8829f04cd531-->
<!--c811ce1c-82ad-49cd-9bbb-32a58c53fd86-->
<!--title-->
## Retry command for [Build#74857](https://buildkite.com/redpanda/redpanda/builds/74857)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":false}
```
<!--/body-->

nguyen-andrew commented:
/ci-repeat 1
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":false}

nguyen-andrew commented:
/cdt
dt-repeat=1
release
tests/rptest/scale_tests/schema_registry_scale_test.py

nguyen-andrew commented:
/dt
dt-repeat=10
release
tests/rptest/scale_tests/schema_registry_scale_test.py

nguyen-andrew commented:
/dt
dt-repeat=1
release
tests/rptest/scale_tests/schema_registry_scale_test.py

nguyen-andrew commented:
[Force push](https://github.com/redpanda-data/redpanda/compare/68ae7c7304dee5f39852ad314107e80172df774c..2d1a114c109d3734555f68ea26d92acef1d4a62e) to adjust the `schema_registry_scale_test.py` scale test classes to inherit from the `SchemaRegistryEndpoints` base class instead of the `SchemaRegistryTestMethods` and `SchemaRegistryModeMutableTest` classes. This is so that CDT only runs the scale tests defined in this file and not the ones associated with `SchemaRegistryTestMethods` and `SchemaRegistryModeMutableTest`

nguyen-andrew commented:
Currently unable to validate the new scale tests with the `/cdt` github command ([issue being investigated here](https://redpandadata.atlassian.net/browse/DEVPROD-3489)), so I've manually validated them using `duck.py` to deploy and run cdt myself. 
@BenPope 
The logs generated during those test runs (before and after the changes) are attached below:
[without_chunked_vec.log](https://github.com/user-attachments/files/23175811/without_chunked_vec.log) (all 5 tests fail due to oversized alloc)
[with_chunked_vec.log](https://github.com/user-attachments/files/23175812/with_chunked_vec.log) (all 5 tests pass)

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/75075-->
<!--7fa460e5-14d2-48e4-ac3c-d1b126440184-->
<!--title-->
## Retry command for [Build#75075](https://buildkite.com/redpanda/redpanda/builds/75075)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[1,"path"],"test_case":{"name":"(TS_Read == True, TS_Timequery == True, SpilloverManifestUploaded == True)"}}
```
<!--/body-->

nguyen-andrew commented:
/ci-repeat 1
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[1,"path"],"test_case":{"name":"(TS_Read == True, TS_Timequery == True, SpilloverManifestUploaded == True)"}}

nguyen-andrew commented:
/ci-repeat 1

vbotbuildovich commented:
/backport v25.2.x

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
Failed to create a backport PR to v25.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-28073-v25.2.x-707 remotes/upstream/v25.2.x
git cherry-pick -x 5079155465 282be03a62 59f0855dfd dce36ea444 5dadb0ef7c 8d61524628 6af58e72b9 c33b98b273 612ed1f098 d069c41708 5749fad31a 27c724907e 0fa51a3237 6f4f6377c7 bc6672840e d3dfe796e4 b7906e69eb cf09113cb4 1cb93c7c3e d542c098ce 6c8a99db95 8749962eb1 2d1a114c10
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/18949757618).

vbotbuildovich commented:
Failed to create a backport PR to v25.1.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-28073-v25.1.x-895 remotes/upstream/v25.1.x
git cherry-pick -x 5079155465 282be03a62 59f0855dfd dce36ea444 5dadb0ef7c 8d61524628 6af58e72b9 c33b98b273 612ed1f098 d069c41708 5749fad31a 27c724907e 0fa51a3237 6f4f6377c7 bc6672840e d3dfe796e4 b7906e69eb cf09113cb4 1cb93c7c3e d542c098ce 6c8a99db95 8749962eb1 2d1a114c10
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/18949757679).
